File: go/problem_51.go
package main

import "math/rand"

func shuffle(deck []int) {
	for i := len(deck) - 1; i > 0; i-- {
		j := rand.Intn(i + 1)
		deck[i], deck[j] = deck[j], deck[i]
	}
}

func problem51() {
	println("Problem 51:")
	deck := make([]int, 52)
	for i := range deck {
		deck[i] = i
	}
	shuffle(deck)
	seen := make(map[int]bool)
	for _, v := range deck {
		if seen[v] {
			panic("duplicate card")
		}
		seen[v] = true
	}
	if len(seen) != 52 {
		panic("wrong count")
	}
	println("All Tests Passed")
}

File: lua/problem_51.lua
local M = {}

function M.shuffle(deck)
    for i = #deck, 2, -1 do
        local j = math.random(1, i)
        deck[i], deck[j] = deck[j], deck[i]
    end
end

function M.problem51()
    print("Problem 51:")
    local deck = {}
    for i = 1, 52 do deck[i] = i end
    M.shuffle(deck)
    local seen = {}
    for _, v in ipairs(deck) do
        assert(not seen[v], "duplicate card")
        seen[v] = true
    end
    assert(#deck == 52)
    print("All Tests Passed")
end

return M
